osquery/events/windows/usn_journal_reader.cpp

```diff
@@ -0,0 +1,879 @@
+  auto& context = d_->journal_reader_context;
+
+  {
+    WriteLock lock(context->processed_records_mutex);
+
+    context->processed_record_list.reserve(
+        context->processed_record_list.size() + record_list.size());
+
+    context->processed_record_list.insert(context->processed_record_list.end(),
+                                          record_list.begin(),
+                                          record_list.end());
+
+    context->processed_records_cv.notify_all();
+  }
+}
+
+void USNJournalReader::start() {
+  // Acquire a handle to the device, as well as the journal id and the initial
+  // sequence number
+  auto status = initialize();
+  if (!status.ok()) {
+    LOG(ERROR) << status.getMessage();
+    return;
+  }
+
+  // Enter the main loop, listening for journal changes
+  while (!interrupted() && !d_->journal_reader_context->terminate) {
+    status = acquireRecords();
+    if (!status.ok()) {
+      LOG(ERROR) << status.getMessage();
+      return;
+    }
+
+    // Process the records
+    std::vector<USNJournalEventRecord> record_list;
+    status = processAcquiredRecords(record_list);
+    if (!status.ok()) {
+      LOG(ERROR) << status.getMessage();
+      return;
+    }
+
+    // Send the new records to the event publisher
+    dispatchEventRecords(record_list);
+  }
+}
+
+void USNJournalReader::stop() {
+  if (d_->volume_handle != INVALID_HANDLE_VALUE) {
+    ::CloseHandle(d_->volume_handle);
+    d_->volume_handle = INVALID_HANDLE_VALUE;
+  }
+}
+
+USNJournalReader::USNJournalReader(
+    USNJournalReaderContextRef journal_reader_context)
+    : InternalRunnable("USNJournalReader"), d_(new PrivateData) {
+  d_->journal_reader_context = journal_reader_context;
+}
+
+USNJournalReader::~USNJournalReader() {}
+
+// TODO: Needs tests; see https://github.com/osquery/osquery/issues/5847
+Status USNJournalReader::DecompressRecord(
+    std::vector<USNJournalEventRecord>& new_records,
+    const USNJournalEventRecord& base_record,
+    DWORD journal_record_reason,
+    USNPerFileLastRecordType& per_file_last_record_type_map) {
+  for (const auto& reason_bit : kUSNChangeReasonFlagList) {
+    if ((journal_record_reason & reason_bit) == 0) {
+      continue;
+    }
+
+    auto new_record = base_record;
+    if (!USNParsers::GetEventType(
+            new_record.type, reason_bit, base_record.attributes)) {
+      return Status(1, "Failed to get the event type");
+    }
+
+    bool emit_record = true;
+    auto last_file_state_it =
+        per_file_last_record_type_map.find(new_record.node_ref_number);
+
+    if (last_file_state_it == per_file_last_record_type_map.end()) {
+      emit_record = true;
+
+    } else {
+      auto last_file_state = last_file_state_it->second;
+      emit_record = last_file_state != new_record.type;
+    }
+
+    if (emit_record) {
+      if (FLAGS_usn_journal_reader_debug) {
+        TLOG << new_record;
+      }
+
+      new_records.push_back(std::move(new_record));
```

Comment:
And why take one incoming record and generate multiple records, one for each reason bit?  Could the `action` column not contain multiple actions that map to the bitfield ? "CREATED|WRITE|DELETE" During a recent time period?

Reply:
> Could the `action` column not contain multiple actions that map to the bitfield ? "CREATED|WRITE|DELETE" During a recent time period?

I think it could, but that might make queries on specific actions a little more annoying (and might make it easy to write subtly incorrect queries, like wanting to query for "CREATED|WRITE" but missing it because of "CREATED|SOMETHING_ELSE|WRITE").